### ISSUE TITLE
AssertionError: TemplatedFile consistency failure when linting SparkSQL file in 4.0.0a1

### ISSUE THREAD

**Issue by noir19** [sqlfluff/sqlfluff#7228]
### Search before asking

- [x] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

Running the linter against a SparkSQL file throws an internal `AssertionError` in `TemplatedFile.__init__()`, so the lint never completes.

### Expected Behaviour

Linting should succeed or return normal rule violations for the file.

### Observed Behaviour

```shell
$ sqlfluff lint -d sparksql src/sql/odps/grade/grade_exec_date_inf.sql
WARNING    Unable to lint src/sql/odps/grade/grade_exec_date_inf.sql due to an internal error. Please report this as an issue with your query's contents and stacktrace below!
Traceback (most recent call last):
  File ".../sqlfluff/core/linter/runner.py", line 116, in run
    yield partial()
  File ".../sqlfluff/core/linter/linter.py", line 809, in lint_rendered
    parsed = cls.parse_rendered(rendered)
  File ".../sqlfluff/core/linter/linter.py", line 336, in parse_rendered
    tokens, lex_errors = cls._lex_templated_file(variant, rendered.config)
  File ".../sqlfluff/core/linter/linter.py", line 183, in _lex_templated_file
    segments, lex_vs = lexer.lex(templated_file)
  File ".../sqlfluff/core/parser/lexer.py", line 939, in lex
    py_template = TemplatedFile(
  File ".../sqlfluff/core/templaters/base.py", line 199, in __init__
    assert rfs.source_idx == pos, (
AssertionError: TemplatedFile. Consistency fail on running source length: 33 != 47
```

### How to reproduce

1. Create a clean environment and install `sqlfluff[rs]==4.0.0a1`.

2.  Use the following SQL statements to create a file called `grade_exec_date_inf.sql`

    ```sql
    INSERT OVERWRITE TABLE :project_name.GRADE_EXEC_DATE_INF PARTITION (DATA_DT = ':data_dt')
    SELECT
        T3.CORP_ORG_CODE,        -- 法人机构代码
        T3.CUST_FLG AS CUST_TYP,
        T3.SCM_ID,
        T3.SCM_ID AS EXE_SCM_ID,
        ':data_dt' AS APP_DT,
        T3.EVAL_RULE_SET,
        T3.STAR_RULE_TYP,
        T3.REL_REL_FLG,
        T3.REL_REL_RTG,
        T3.VRSN_STAT,
        T3.SVC_GRD_DFR_FLG,
        T3.SVC_GRD_DFR_PRD_NUM,
        T3.REL_GRD_DFR_FLG,
        T3.REL_GRD_DFR_PRD_NUM,
        T3.BUFF_PRD,
        T3.NEW_CUST_INL_RTG,
        T3.NEW_CUST_VAL_INL_RTG,
        CASE
            WHEN T3.EVAL_FREQ = '0' THEN T2.DATE_MEND
            WHEN T3.EVAL_FREQ = '1' THEN T2.DATE_QEND
            WHEN T3.EVAL_FREQ = '2' THEN T2.DATE_YEND
            ELSE ''
        END AS EXE_DT,
        T3.EVAL_START_DT,
        T3.EVAL_END_DT
    FROM (
        SELECT
            ':data_dt' AS CUR_DT,
            A.INSTU_CDE AS CORP_ORG_CODE
        FROM :project_name.RE_ADMIN_SM_INSTU A
    ) T1
    INNER JOIN :project_name.ADMIN_DATE_DIM T2
        ON T1.CUR_DT = T2.DATA_DT
    INNER JOIN :project_name.RE_GRADE_SCM_MGT_INF T3
        ON T1.CORP_ORG_CODE = T3.CORP_ORG_CODE
    WHERE
        T3.ENABLE_FLG = '1'
        AND T3.VRSN_STAT IN ('0', '1')
        AND T3.CORP_ORG_CODE <> '0100'
        AND T3.DEL_FLG = '0';
    ```

3. Run:

   ```bash
   sqlfluff lint -d sparksql grade_exec_date_inf.sql
   ```

### Additional Note

When using sqlfluff==3.5.0, the command runs successfully without any issues.
However, under sqlfluff[rs]==4.0.0a1, the linting process reports errors only when the SQL file contains Chinese comments.
After removing the Chinese characters or comments, the same command executes normally.

Therefore, it appears that the issue is related to the presence of Chinese characters in the SQL file.

### Dialect

sparksql

### Version

sqlfluff[rs]==4.0.0a1

### Configuration

```yaml
[sqlfluff]
large_file_skip_byte_limit = 0
templater = placeholder
exclude_rules = ST05, AL01, RF01
max_line_length = 120

[sqlfluff:templater:placeholder]
param_style = colon
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by alanmcruickshank** [comment]
Firstly, thankyou so much for testing the alpha release! ❤️ 

@keraion, unless this is super easy for you, I might see if I can solve it - I'm going to need to learn the rust codebase at some point!

**Comment by keraion** [comment]
@alanmcruickshank feel free to take a look at it. My guess is that it will be in the python mod for the `PyTemplatedFile`, likely a usage of `.len()` instead of `.chars().count()` for calculating codepoints.

**Comment by alanmcruickshank** [comment]
I've narrowed it down to this query triggering the issue:

```sql
SELECT
    d, -- 法
    ':c' AS f
```

oddly, it's not just the unicode characters, and it's not just the param style. I think there's an interaction between the two in some way.

**Comment by keraion** [comment]
That makes sense as a non-templated version will only have a single raw file slice which only tracks the starting index.
